src/coreclr/nativeaot/BuildIntegration/Microsoft.NETCore.Native.Unix.targets

```diff
@@ -222,6 +221,7 @@ The .NET Foundation licenses this file to you under the MIT license.
       <NativeSystemLibrary Include="log" Condition="'$(_linuxLibcFlavor)' == 'bionic'" />
       <NativeSystemLibrary Include="icucore" Condition="'$(_IsApplePlatform)' == 'true'" />
       <NativeSystemLibrary Include="m" />
+      <NativeSystemLibrary Include="atomic" Condition="'$(_targetArchitecture)' == 'riscv64'" />
```

Comment:
I assume that this is to resolve `__atomic_compare_exchange_16` missing symbol. Does riscv64 have proper 16-byte atomic operation or is this hiding the same reliability bug as https://github.com/dotnet/runtime/issues/109276 ?

Reply:
It requires zacas extension from under development RVA23 profile.

Reply:
`zacas` extension was ratified recently for RVA23 U64 profile and available in kernel 6.8: https://lf-rise.atlassian.net/wiki/spaces/HOME/pages/8587962/RVA23+Profile. I'm on RVA22 machine with kernel 6.6, so this is necessary right now; a C program with __sync_val_compare_and_swap(__int128_t..) requires libatomic. The disassembly (https://gist.github.com/am11/a884ba0583b5dd4abd3e231f0b8989eb) suggests that it's not using locks, but iterative way to implement CAS-128 with smaller CAS-64 to maintain the semantics.

I think this is also future-proof since libatomic will handle the hardware vs. software fallback selection for us.

Reply:
`zacas` extension was ratified recently for RVA23 U64 profile and available in kernel 6.8: https://lf-rise.atlassian.net/wiki/spaces/HOME/pages/8587962/RVA23+Profile. I'm on RVA22 machine with kernel 6.6, so this is necessary right now; a C program with __sync_val_compare_and_swap(__int128_t..) requires libatomic. The disassembly (https://gist.github.com/am11/a884ba0583b5dd4abd3e231f0b8989eb) suggests that it's not using locks, but iterative way to implement CAS-128 with smaller CAS-64 to maintain the semantics. It can be a bit slower when retry is imminent.

I think this is also future-proof since libatomic will handle the hardware vs. software fallback selection depending on the underlying machine.

Reply:
> iterative way to implement CAS-128 with smaller CAS-64 to maintain the semantics.

I do not think that this is maintains semantics required by the runtime. The current runtime implementation requires proper CAS-128.

Reply:
IIUC, LoongArch libatomic impl was using locks, riscv's impl. does not use locks. If there is a better alternative available, I can switch to that. What I understood from community is libatomic is maintaining the semantics on riscv64 and it can be slower. RVA23 is not going to get completed anytime soon.

BTW, when we will cross that bridge then we can revisit it. Currently it's a stopgap and we have other work to complete the implementation. After that, we can stress test it to find the issue it causes.

Reply:
The disassembly that you have shared has two regular 64-bit stores: https://gist.github.com/am11/a884ba0583b5dd4abd3e231f0b8989eb#file-__atomic_compare_exchange_16-disasm-s-L26-L27 . I do not see how one can do 128-bit CAS with two 64-bit regular stores. Am I missing some trick?

I am fine with a temporary workaround, but it needs to be documented as a workaround - TODOs with links to issues.

Reply:
How does NativeAOT work on X64 considering that DCAS isn't guaranteed there? Does it have an additional requirement for that?

Reply:
I will add a TODO.

What I know is Debian has a very high bar of promoting an ISA to officially supported status. It took them multiple years to port all software packages to riscv64gc and given it an official status (and Ubuntu 24.04 added its official support as well). So is Fedora, Gentoo and Alpine. Now if libatomic implementation from core gcc toolchain is not doing what it is supposed it, while possible, it sounds highly unlikely.

This PR brings PublishAot to succeed and running the published app on device (and qemu) is throwing SIGILL from managed->native internal call (native->manage calls are working), which suggests we have more work to do. I will share details in the issue after this. Lets get the hello world and some basic stuff working.

Reply:
> How does NativeAOT work on X64 considering that DCAS isn't guaranteed there?

NativeAOT required 16-byte compare-exchange on x64 today: https://github.com/dotnet/runtime/blob/db7e9abb70208a67e686fdb6a593618649a2ed7e/src/coreclr/nativeaot/CMakeLists.txt#L26-L27 . It won't work without it.